### ISSUE TITLE
Request for the implementation of GraphTSR

### ISSUE THREAD

**Issue by huntzhan** [Academic-Hammer/SciTSR#7]
Dear Authors,

Thanks for your great work.

In the paper "Complicated Table Structure Recognition" you propose a novel graph neural network for recognizing the table structure in PDF files, named GraphTSR. I've tried to reproduce the experiment but the results do not align with the numbers presented in the paper. I am currently working on TSR and would like to ask if you could send me the implementation for further analysis as it would help me verify your results. Please let me know if you have concerns about opening sourcing the code.

My email: huntzhan.dev@gmail.com

Best wishes,

Hunt Zhan

**Comment by CZWin32768** [comment]
The code of GraphTSR is a part of a larger closed-source system. We have stripped out some parts of the code to this repo, and we are also willing to release all of the code. We will try to release the open-source version code as early as possible.

Just feel free to ask me any questions about the implementation details. Sorry for the inconvenience.

**Comment by huntzhan** [comment]
Thanks for the response!

Could you help confirm the details of your model structure? In the paper, the dropout is applied *with p = 0.4 to the output of each sub-layer before it is added to the sublayer input*. From my understanding the model looks like this:

```
# Input 
H_v = Linear( vertex features )
H_e = Linear( edge features )


# Attention Block (only presents the edge-to-vertex attention block)
Q = H_v x W_Q
K = H_e x W_K
V = H_e x W_V

// DROPOUT!
// Note: dropout is NOT applied inside the GraphAtt, i.e. to the the attention coefficients
H˜_v = Norm( H_v + Dropout( GraphAtt( Q, K, V ) ) ) 

// DROPOUT!
// Note: FFN( H˜_v ) = Linear_1 ( RELU ( Linear_2 ( H˜_v ) ) ),
// and dropout is NOT applied inside the FFN.
H_v = Norm( H˜_v + Dropout( FFN( H˜_v ) ) )


// Output
Hypothesis_e = Softmax ( Linear ( H_e ) )
```

Questions:
1. Dropout is only to the results of `GraphAtt` and `FFN` in the attention block, meaning that dropout is neither applied inside the `GraphAtt` and `FFN` nor applied to the input & output linear transformations,  is that correct?
2. `FFN` consists of two weights `W_1` and `W_2`, what your choice of the shapes?
3. Both the input and output linear transformations are one-layer linear transformation without non-linear activation, in contract to the `FFN`, is that correct?
4. Does the model structure described above match your implementation?

**Comment by CZWin32768** [comment]
Sorry for the delayed release of the code. You can find the details of the model structure at [here](https://github.com/Academic-Hammer/SciTSR/blob/master/scitsr/model.py).

**Comment by huntzhan** [comment]
Thank you! The implementation you provided has resolved all of my questions.

**Comment by huntzhan** [comment]
@CZWin32768 In the paper 
> We use the same pre-processing procedure as (Shigarov et al., 2016), and employ a *simple post-processing to convert a labeled graph to structure data*.

BTW, could you help briefly describe how to convert a labeled graph into the final table structure? I think this task could be tricky since there could be some noisy edges in the graph.

**Comment by sindhurk** [comment]
@huntzhan 
Good to know that you could reproduce and train this model. Could you help share your code to input the training data and train the model

email: [sindhuraman610@gmail.com](url)

Thanks

**Comment by AtulKumar4** [comment]
> The code of GraphTSR is a part of a larger closed-source system. We have stripped out some parts of the code to this repo, and we are also willing to release all of the code. We will try to release the open-source version code as early as possible.
> 
> Just feel free to ask me any questions about the implementation details. Sorry for the inconvenience.

**Comment by AtulKumar4** [comment]
@CZWin32768 Could you please help me with the inference part on new data

**Comment by AtulKumar4** [comment]
@huntzhan Could you help me with the inference part if you have already figured it out?
